classifier.py

```diff
@@ -328,8 +342,52 @@ def smart_loss(model: nn.Module, b_ids: torch.Tensor, b_mask: torch.Tensor, orgi
         embeddings_perturbed = embeddings_perturbed + args.eta * grad
         # Project embeddings_perturbed back to the L_inf ball of radius epsilon centered at start_embeddings.
         embeddings_perturbed = start_embeddings + \
-            torch.clamp(embeddings_perturbed - start_embeddings, -args.epsilon, args.epsilon)
-    return loss_perturbed
+            torch.clamp(embeddings_perturbed - start_embeddings, -
+                        args.epsilon, args.epsilon)
+
+    # Calculating one more time for the final perturbatin loss, after we find
+    # the most adversarial perturbation.
+    logits = model(embeddings_perturbed, b_mask, is_embedding=True)
+
+    return F.kl_div(
+        F.log_softmax(logits, dim=-1),
+        F.log_softmax(orginal_logits, dim=-1),
+        reduction='batchmean',
+        log_target=True,
+    ) + F.kl_div(
+        F.log_softmax(orginal_logits, dim=-1),
+        F.log_softmax(logits, dim=-1),
+        reduction='batchmean',
+        log_target=True
+    )
+
+
+def get_bregmman_loss(model_tilde: nn.Module, logits: torch.Tensor, b_ids: torch.Tensor, b_mask: torch.Tensor):
+    # TODO: unify the l_s calculation into a single function that also usable for regression task.
+    # Disable grad as we never going to update logits_tilde.
+    with torch.no_grad():
+        logits_tilde = model_tilde(b_ids, b_mask)
+
+    bregmman_loss = F.kl_div(
+        F.log_softmax(logits, dim=-1),
+        F.log_softmax(logits_tilde, dim=-1),
+        reduction='batchmean',
+        log_target=True,
+    ) + F.kl_div(
+        F.log_softmax(logits_tilde, dim=-1),
+        F.log_softmax(logits, dim=-1),
+        reduction='batchmean',
+        log_target=True
+    )
+
+    return bregmman_loss
+
+
+def update_model_tilde(model_tilde: nn.Module, model: nn.Module, beta: float):
+    with torch.no_grad():
+        for param_tilde, param_update in zip(model_tilde.parameters(), model.parameters()):
+            param_tilde.mul_(1 - beta)
```

Comment:
shouldn't this be `param_tilde.mul_(beta)`
and next line be `param_tilde.add_(param_update, 1 - beta)`?